[symint] Make TensorImpl::sizes_and_strides_ contain SymInt

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #77795
* #77914

Change our representation of sizes and strides to contain SymInts
instead of int64_t.

Right now it's not actually possible to create a Tensor with symbolic
shape, so this change is intended to be a no-op.

But the intended behavior is:
- If you create a Tensor with symbolic shape, a `CustomSizes` policy
will be set, and the `has_symbolic_sizes_strides_` bit will be set. (not
currently implemented)
- Calling any TensorImpl function that naively interacts with sizes and
strides will throw. For hot-path functions (`sizes()`, `strides()`), we
make use of the existing policy check to throw. For others, we just have
a regular `TORCH_CHECK(!has_symbolic_sizes_strides_)`.

This also undoes the explicit constructor I made in
https://github.com/pytorch/pytorch/pull/77666; it ended up being more
annoying than useful when making these changes.

Differential Revision: [D36520548](https://our.internmc.facebook.com/intern/diff/D36520548)

suo committed:
[wip] Make TensorImpl::sizes_and_strides_ contain SymInt

Change our representation of sizes and strides to contain SymInts
instead of int64_t.

Right now it's not actually possible to create a Tensor with symbolic
shape, so this change is intended to be a no-op.

But the intended behavior is:
- If you create a Tensor with symbolic shape, a `CustomSizes` policy
will be set, and the `has_symbolic_sizes_strides_` bit will be set. (not
currently implemented)
- Calling any TensorImpl function that naively interacts with sizes and
strides will throw. For hot-path functions (`sizes()`, `strides()`), we
make use of the existing policy check to throw. For others, we just have
a regular `TORCH_CHECK(!has_symbolic_sizes_strides_)`.

[ghstack-poisoned]

suo committed:
Update on "[wip] Make TensorImpl::sizes_and_strides_ contain SymInt"


Change our representation of sizes and strides to contain SymInts
instead of int64_t.

Right now it's not actually possible to create a Tensor with symbolic
shape, so this change is intended to be a no-op.

But the intended behavior is:
- If you create a Tensor with symbolic shape, a `CustomSizes` policy
will be set, and the `has_symbolic_sizes_strides_` bit will be set. (not
currently implemented)
- Calling any TensorImpl function that naively interacts with sizes and
strides will throw. For hot-path functions (`sizes()`, `strides()`), we
make use of the existing policy check to throw. For others, we just have
a regular `TORCH_CHECK(!has_symbolic_sizes_strides_)`.

[ghstack-poisoned]

suo committed:
Update on "[wip] Make TensorImpl::sizes_and_strides_ contain SymInt"


Change our representation of sizes and strides to contain SymInts
instead of int64_t.

Right now it's not actually possible to create a Tensor with symbolic
shape, so this change is intended to be a no-op.

But the intended behavior is:
- If you create a Tensor with symbolic shape, a `CustomSizes` policy
will be set, and the `has_symbolic_sizes_strides_` bit will be set. (not
currently implemented)
- Calling any TensorImpl function that naively interacts with sizes and
strides will throw. For hot-path functions (`sizes()`, `strides()`), we
make use of the existing policy check to throw. For others, we just have
a regular `TORCH_CHECK(!has_symbolic_sizes_strides_)`.

[ghstack-poisoned]

suo committed:
Update on "[wip] Make TensorImpl::sizes_and_strides_ contain SymInt"


Change our representation of sizes and strides to contain SymInts
instead of int64_t.

Right now it's not actually possible to create a Tensor with symbolic
shape, so this change is intended to be a no-op.

But the intended behavior is:
- If you create a Tensor with symbolic shape, a `CustomSizes` policy
will be set, and the `has_symbolic_sizes_strides_` bit will be set. (not
currently implemented)
- Calling any TensorImpl function that naively interacts with sizes and
strides will throw. For hot-path functions (`sizes()`, `strides()`), we
make use of the existing policy check to throw. For others, we just have
a regular `TORCH_CHECK(!has_symbolic_sizes_strides_)`.

[ghstack-poisoned]

suo committed:
Update on "[wip] Make TensorImpl::sizes_and_strides_ contain SymInt"


Change our representation of sizes and strides to contain SymInts
instead of int64_t.

Right now it's not actually possible to create a Tensor with symbolic
shape, so this change is intended to be a no-op.

But the intended behavior is:
- If you create a Tensor with symbolic shape, a `CustomSizes` policy
will be set, and the `has_symbolic_sizes_strides_` bit will be set. (not
currently implemented)
- Calling any TensorImpl function that naively interacts with sizes and
strides will throw. For hot-path functions (`sizes()`, `strides()`), we
make use of the existing policy check to throw. For others, we just have
a regular `TORCH_CHECK(!has_symbolic_sizes_strides_)`.

[ghstack-poisoned]

suo committed:
Update on "[symint] Make TensorImpl::sizes_and_strides_ contain SymInt"


Change our representation of sizes and strides to contain SymInts
instead of int64_t.

Right now it's not actually possible to create a Tensor with symbolic
shape, so this change is intended to be a no-op.

But the intended behavior is:
- If you create a Tensor with symbolic shape, a `CustomSizes` policy
will be set, and the `has_symbolic_sizes_strides_` bit will be set. (not
currently implemented)
- Calling any TensorImpl function that naively interacts with sizes and
strides will throw. For hot-path functions (`sizes()`, `strides()`), we
make use of the existing policy check to throw. For others, we just have
a regular `TORCH_CHECK(!has_symbolic_sizes_strides_)`.

This also undoes the explicit constructor I made in
https://github.com/pytorch/pytorch/pull/77666; it ended up being more
annoying than useful when making these changes.

[ghstack-poisoned]

suo commented:
@suo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36520548).

ezyang commented on aten/src/ATen/OpaqueTensorImpl.h:
```diff
@@ -70,7 +70,11 @@ struct TORCH_API OpaqueTensorImpl : public TensorImpl {
+        dtype(),
+        device(),
+        opaque_handle_,
+        expectIntArrayRef(sizes_and_strides_.sizes_arrayref()));
```
If there is an invariant that OpaqueTensorImpl never has symbolic ints this should be an unchecked cast, but this path is not perf critical so it seems ok

ezyang commented on c10/core/SymInt.h:
```diff
@@ -47,12 +48,46 @@ class C10_API SymInt {
+  }
+
+  SymInt operator*(int64_t sci) const {
+    TORCH_CHECK(!this->is_symbolic(), "Symbolic mul isn't supported yet");
```
We are going to want this out-of-line in a cpp file so we aren't inlining a big blob of exception throwing code everywhere.

ezyang commented on c10/core/SymIntArrayRef.h:
```diff
@@ -73,6 +73,16 @@ class SymIntArrayRef final {
+      TORCH_INTERNAL_ASSERT_DEBUG_ONLY(
+          array_ref[i] >= 0,
+          "Expected all elements in array to be non-negative");
+    }
```
to be adjusted after our negative numbers discussion

ezyang commented on c10/core/impl/SizesAndStrides.h:
```diff
@@ -111,15 +113,31 @@ class C10_API SizesAndStrides {
+    }
+
+    return false;
+  }
```
Maybe call this `slow_has_sym` to deter people from calling it

ezyang commented on c10/core/SymInt.h:
```diff
@@ -47,12 +48,46 @@ class C10_API SymInt {
+  }
 
   std::shared_ptr<SymbolicIntNode> toSymbolicIntNode();
   static c10::SymInt toSymInt(std::shared_ptr<SymbolicIntNode> sin);
```
re data below, it would be nice to not directly use data(), but name it something scary like unchecked_expect_int() to note that there is a precondition that it must not be symbolic. I wish I could dump data too but IValue would not work.

ezyang commented on c10/core/TensorImpl.cpp:
```diff
@@ -584,26 +584,31 @@ void TensorImpl::Extend(int64_t num, float growthPct) {
-  SizesVector newDims(
-      sizes_and_strides_.sizes_begin(), sizes_and_strides_.sizes_end());
+  IntArrayRef sizes_and_strides =
+      expectIntArrayRef(sizes_and_strides_.sizes_arrayref());
```
expectIntArrayRef should be renamed to have "slow" in its name. Because you have tested `has_symbolic_sizes_strides_` an unchecked cast is OK, and we should have a way of spelling that too.

ezyang reviewed (approved):
LGTM

suo committed:
Update on "[symint] Make TensorImpl::sizes_and_strides_ contain SymInt"


Change our representation of sizes and strides to contain SymInts
instead of int64_t.

Right now it's not actually possible to create a Tensor with symbolic
shape, so this change is intended to be a no-op.

But the intended behavior is:
- If you create a Tensor with symbolic shape, a `CustomSizes` policy
will be set, and the `has_symbolic_sizes_strides_` bit will be set. (not
currently implemented)
- Calling any TensorImpl function that naively interacts with sizes and
strides will throw. For hot-path functions (`sizes()`, `strides()`), we
make use of the existing policy check to throw. For others, we just have
a regular `TORCH_CHECK(!has_symbolic_sizes_strides_)`.

This also undoes the explicit constructor I made in
https://github.com/pytorch/pytorch/pull/77666; it ended up being more
annoying than useful when making these changes.

Differential Revision: [D36520548](https://our.internmc.facebook.com/intern/diff/D36520548)

[ghstack-poisoned]

suo committed:
Update on "[symint] Make TensorImpl::sizes_and_strides_ contain SymInt"


Change our representation of sizes and strides to contain SymInts
instead of int64_t.

Right now it's not actually possible to create a Tensor with symbolic
shape, so this change is intended to be a no-op.

But the intended behavior is:
- If you create a Tensor with symbolic shape, a `CustomSizes` policy
will be set, and the `has_symbolic_sizes_strides_` bit will be set. (not
currently implemented)
- Calling any TensorImpl function that naively interacts with sizes and
strides will throw. For hot-path functions (`sizes()`, `strides()`), we
make use of the existing policy check to throw. For others, we just have
a regular `TORCH_CHECK(!has_symbolic_sizes_strides_)`.

This also undoes the explicit constructor I made in
https://github.com/pytorch/pytorch/pull/77666; it ended up being more
annoying than useful when making these changes.

Differential Revision: [D36520548](https://our.internmc.facebook.com/intern/diff/D36520548)

[ghstack-poisoned]

suo committed:
Update on "[symint] Make TensorImpl::sizes_and_strides_ contain SymInt"


Change our representation of sizes and strides to contain SymInts
instead of int64_t.

Right now it's not actually possible to create a Tensor with symbolic
shape, so this change is intended to be a no-op.

But the intended behavior is:
- If you create a Tensor with symbolic shape, a `CustomSizes` policy
will be set, and the `has_symbolic_sizes_strides_` bit will be set. (not
currently implemented)
- Calling any TensorImpl function that naively interacts with sizes and
strides will throw. For hot-path functions (`sizes()`, `strides()`), we
make use of the existing policy check to throw. For others, we just have
a regular `TORCH_CHECK(!has_symbolic_sizes_strides_)`.

This also undoes the explicit constructor I made in
https://github.com/pytorch/pytorch/pull/77666; it ended up being more
annoying than useful when making these changes.

Differential Revision: [D36520548](https://our.internmc.facebook.com/intern/diff/D36520548)

[ghstack-poisoned]

suo commented:
@pytorchbot merge

suo commented:
@pytorchbot merge

suo commented:
@pytorchbot merge -f

suo committed:
Update on "[symint] Make TensorImpl::sizes_and_strides_ contain SymInt"


Change our representation of sizes and strides to contain SymInts
instead of int64_t.

Right now it's not actually possible to create a Tensor with symbolic
shape, so this change is intended to be a no-op.

But the intended behavior is:
- If you create a Tensor with symbolic shape, a `CustomSizes` policy
will be set, and the `has_symbolic_sizes_strides_` bit will be set. (not
currently implemented)
- Calling any TensorImpl function that naively interacts with sizes and
strides will throw. For hot-path functions (`sizes()`, `strides()`), we
make use of the existing policy check to throw. For others, we just have
a regular `TORCH_CHECK(!has_symbolic_sizes_strides_)`.

This also undoes the explicit constructor I made in
https://github.com/pytorch/pytorch/pull/77666; it ended up being more
annoying than useful when making these changes.

Differential Revision: [D36520548](https://our.internmc.facebook.com/intern/diff/D36520548)

[ghstack-poisoned]

suo commented:
@pytorchbot merge -g
